src/resolve/resolved-dns-answer.c

```diff
@@ -181,11 +181,19 @@ int dns_answer_add(
 
         exist = ordered_set_get(a->items, &tmp);
         if (exist) {
-                /* There's already an RR of the same RRset in place! Let's see if the TTLs more or less
-                 * match. We don't really care if they match precisely, but we do care whether one is 0 and
-                 * the other is not. See RFC 2181, Section 5.2. */
-                if ((rr->ttl == 0) != (exist->rr->ttl == 0))
-                        return -EINVAL;
+                /* There's already an RR of the same RRset in place! Let's see if the TTLs more or
+                 * less match. RFC 2181, Section 5.2 suggests clients should reject RRsets
+                 * containing RRs with differing TTLs. We are more tolerant of this situation except
+                 * if one RR has a zero TTL and the other a nonzero TTL. In mDNS, zero TTLs are
+                 * special, so we must error in that case. */
+                if ((rr->ttl == 0) != (exist->rr->ttl == 0)) {
+                        log_debug("Attempting to merge RRs with zero TTL and nonzero TTL: %s %s",
+                                  dns_resource_record_to_string(rr),
+                                  dns_resource_record_to_string(exist->rr));
+                        if (exist->flags & DNS_ANSWER_MDNS || flags & DNS_ANSWER_MDNS) {
+                                return -EINVAL;
+                        }
```

Comment:
coding style, no {} around single line if block.

also, please add () around the & expressions, it's otherwise just too weird to read...

Also, in the log message please say what you are doing, i.e. whether you refuse or if you allow